src/main/java/ee/carlrobert/llm/client/watsonx/WatsonxAuthenticator.java

```diff
@@ -0,0 +1,128 @@
+    this.client = new OkHttpClient().newBuilder()
+        .build();
+    MediaType mediaType = MediaType.parse("application/x-www-form-urlencoded");
+    RequestBody body = RequestBody.create(mediaType,
+        "grant_type=urn:ibm:params:oauth:grant-type:apikey&apikey=" + apiKey);
+    this.request = new Request.Builder()
+        .url("https://iam.cloud.ibm.com/identity/token")
+        .method("POST", body)
+        .addHeader("Content-Type", "application/x-www-form-urlencoded")
+        .build();
+    try {
+      Response response = client.newCall(request).execute();
+      this.bearerToken = OBJECT_MAPPER.readValue(response.body().string(),
+          IBMAuthBearerToken.class);
+    } catch (IOException e) {
+      System.out.println(e);
+    }
+  }
+
+  // Zen API Key
+  public WatsonxAuthenticator(String username, String zenApiKey) {
+    IBMAuthBearerToken token = new IBMAuthBearerToken();
+    String tokenStr = Base64.getEncoder().encode((username + ":" + zenApiKey).getBytes())
+        .toString();
+    token.setAccessToken(tokenStr);
+    this.bearerToken = token;
+    this.isZenApiKey = true;
+  }
+
+  // Watsonx API Key
+  public WatsonxAuthenticator(String username, String apiKey,
+      String host) { // TODO add support for password
+    this.client = new OkHttpClient().newBuilder()
+        .build();
+    ObjectMapper mapper = new ObjectMapper();
+    Map<String, String> authParams = new LinkedHashMap<>();
+    authParams.put("username", username);
+    authParams.put("api_key", apiKey);
+
+    String authParamsStr = "";
+    try {
+      authParamsStr = mapper.writeValueAsString(authParams);
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+
+    MediaType mediaType = MediaType.parse("application/json");
+    RequestBody body = RequestBody.create(mediaType, authParamsStr);
+    // TODO add support for IAM endpoint v1/auth/identitytoken
+    this.request = new Request.Builder()
+        .url(host
+            + "/icp4d-api/v1/authorize")
+        .method("POST", body)
+        .addHeader("Content-Type", "application/json")
+        .build();
+
+    this.expiryRequest = new Request.Builder()
+        .url(host + "/usermgmt/v1/user/tokenExpiry")
+        .addHeader("Accept", "application/json")
+        .addHeader("Authorization", "Bearer " + this.bearerToken.getAccessToken())
+        .build();
+
+    try {
+      Response response = client.newCall(request).execute();
+      this.bearerToken = OBJECT_MAPPER.readValue(response.body().string(),
+          IBMAuthBearerToken.class);
+
+      Response expiry = client.newCall(request).execute();
+      this.bearerToken.setExpiration(
+          OBJECT_MAPPER.readValue(expiry.body().string(), IBMAuthTokenExpiry.class)
+              .getExpiry());
+
+    } catch (IOException e) {
+      System.out.println(e);
+    }
+  }
+
+  private void generateNewBearerToken() {
+    try {
+      Response response = client.newCall(request).execute();
```

Comment:
Resources need to be cleaned up - `try (Response response = client.newCall(request).execute())`